[STORY-989] feat(pg): prepare release postgresql 16 and PostgreSQL Anonymizer (#2969)
Scalingo/documentation: src/_posts/databases/postgresql/2000-01-01-postgresql-anonymizer.md

This extension is particularly useful for organizations handling sensitive data and is designed with **anonymization by design** principles to ensure compliance with privacy regulations such as GDPR, HIPAA, and CCPA.

This extension adds several features and functions. You can find more information on the [official documentation](https://postgresql-anonymizer.readthedocs.io/en/stable/).

## Enabling PostgreSQL® Anonymizer

1. [Create a new user with read only abilities]({% post_url databases/postgresql/2000-01-01-managing-users %})
2. Contact our support team and provide the name of the user you just created to request activation of the extension
3. Our team activates the extension and sets up masking for the given user
4. Once done, you can define and manage masking rules for your user autonomously as described below.

## Example: Dynamic Masking for Read-Only Users

Suppose you have a table named users containing the columns `first_name`, `last_name`, `email`, and `phone` and you want to hide sensitive data dynamically for specific users.

> These users must be `read-only` and should have been masked as described above. You can then declare `SECURITY LABEL` with your regular user:

Reviewer: ```suggestion
These users must be restricted to read-only abilities and must have been masked as described above. You can then declare `SECURITY LABEL` with your regular user.
```

> * Regular roles will still access the original data
> * Masked user will see altered data according to the rules you have set

Reviewer: ```suggestion
* Regular roles still have access to the original data
* Masked users see altered data according to the masking rules
```

Connect as the regular user to the database:

```sql
SELECT * FROM people;

Result:
  id | firstname | lastname |   phone    |         email
 ----+-----------+----------+------------+-------------------------
   1 | Sarah     | Conor    | 0606060606 |  sarah.conor@example.com
   2 | John      | Doe      | 0707070707 |  john.doe@example.com

-- Declare the masking rules
SECURITY LABEL FOR anon ON COLUMN people.lastname
IS 'MASKED WITH FUNCTION anon.dummy_last_name();';

SECURITY LABEL FOR anon ON COLUMN people.phone
IS 'MASKED WITH FUNCTION anon.partial(phone, 2, $$******$$, 2)';
```

> Connect as the "read-only" user to the database and view data:

Reviewer: ```suggestion
Connect as the read-only user to the database and view data:
```

```sql
SELECT * FROM people;

 Result:
  id | firstname | lastname  |    phone     |           email
 ----+-----------+-----------+--------------+----------------------------

>    1 |  Sarah    | Preston   | 06********06 | rayamanda@example.net
>    2 |  John     | Cristian  | 07********07 | mariahpatrick@example.org

Reviewer: ```suggestion
   1 |  Sarah    | Preston   | 06********06 | sarah.conor@example.com
   2 |  John     | Cristian  | 07********07 | john.doe@example.com
```

---
title: PostgreSQL Anonymizer
nav: PostgreSQL Anonymizer
modified_at: 2025-01-24 09:00:00
tags: databases postgresql anonymization extensions
index: 16
---

**PostgreSQL Anonymizer** is an open-source extension for PostgreSQL® that facilitates the masking or replacement of personally identifiable information (PII) and commercially sensitive data directly within a database.

This extension is particularly useful for organizations handling sensitive data and is designed with **anonymization by design** principles to ensure compliance with privacy regulations such as GDPR, HIPAA, and CCPA.

> The extension supports the following features:
> 
> - Static anonymization: Irreversibly anonymize data for development or testing environments.
> - Dynamic masking: Apply real-time masking to hide sensitive information for specific users or roles.
> - Customizable anonymization rules: Use pre-defined or user-defined strategies for anonymization.
> - Multiple anonymization techniques: Including randomization, hashing, and data redaction.
> - Role-based access control: Restrict visibility of sensitive data based on user roles.
> - Compatibility with standard PostgreSQL queries and tools.

Reviewer: Since we don't detail features for the other extensions, I suggest to remove the list and let people check an exhaustive one on the official website:

```suggestion
This extension adds several features and functions. You can find more information on the [official documentation](https://postgresql-anonymizer.readthedocs.io/en/stable/).
```

Example: Hiding Data for a "Masked" User

```sql
-- View data as a regular user
SELECT * FROM people;

Result:
  id | firstname | lastname |   phone    |         email
 ----+-----------+----------+------------+-------------------------
   1 | Sarah     | Conor    | 0606060606 |  sarah.conor@example.com
   2 | John      | Doe      | 0707070707 |  john.doe@example.com

-- Declare the masking rules
SECURITY LABEL FOR anon ON COLUMN people.lastname
IS 'MASKED WITH FUNCTION anon.dummy_last_name();';

SECURITY LABEL FOR anon ON COLUMN people.phone
IS 'MASKED WITH FUNCTION anon.partial(phone, 2, $$******$$, 2)';

-- Connect as the masked user and view data
\c - my_read_only_user
SELECT * FROM people;

 Result:
  id | firstname | lastname  |    phone     |           email            
 ----+-----------+-----------+--------------+----------------------------
   1 |  Reginald | Preston   | +33477420668 | rayamanda@example.net
   2 |  Shaffer  | Cristian  | +33225352636 | mariahpatrick@example.org

> ```

Reviewer: IMHO, this needs to be re-written:

- We start an example on line 50 and then put an `Example:` emphasis on line 52.
- We never explain that the "read-only" user has the "MASKED" role.
- We don't clearly explain how to create this role.
- We never link the role to the user:
  ```
  CREATE ROLE my_read_only_user LOGIN;
  SECURITY LABEL FOR anon ON ROLE my_read_only_user IS 'MASKED';
  GRANT pg_read_all_data to my_read_only_user;
  ```
- The "read-only" user becomes "my_read_only_user" :-/
- The result doesn't look like what's expected:
  - `firstname` has been updated with dummy values, it shouldn't have
  - `phone` is not masked, but replaced with other values
  - `email` has been updated with dummy values, it shouldn't have

Reviewer 2: Updated, please tell me how you like it

> ## Enabling PostgreSQL Anonymizer
> 
> To enable PostgreSQL Anonymizer:

Reviewer: ```suggestion
## Enabling PostgreSQL® Anonymizer

To enable PostgreSQL® Anonymizer:
```

> To enable PostgreSQL Anonymizer:

Reviewer: Can't we automate this? :disappointed:

Reviewer 2: We could but it has been decided we wouldn't do day 1.

Reason is that SUPERUSER is required to initialize the extension, so user is not autonomous handling this.

Reviewer: Arf, I see, this is what I feared.

> 1. Create a "read_only" user through your database dashboard.

Reviewer: ```suggestion
1. [Create a new user with read only abilities]({% post_url databases/postgresql/2000-01-01-managing-users %})
```

> 2. Contact our support team and provide the name of your "read_only" user to request activation of the extension.

Reviewer: ```suggestion
2. Contact our support team and provide the name of the user you just created to request activation of the extension
```

> 3. Our team will activate the extension and set up masking for your "read_only" user.

Reviewer: ```suggestion
3. Our team activates the extension and sets up masking for the given user
```

> 4. Once completed, you can use the extension to define and manage masking rules for your "read_only" user independently.

Reviewer: ```suggestion
4. Once done, you can define and manage masking rules for your user autonomously
```
